packages/mobile/src/qrcode/utils.ts

```diff
@@ -48,22 +54,59 @@ export async function shareSVGImage(svg: SVG) {
 export function* handleBarcode(
   barcode: QrCode,
   addressToE164Number: AddressToE164NumberType,
-  recipientCache: NumberToRecipient
+  recipientCache: NumberToRecipient,
+  e164NumberToAddress: E164NumberToAddressType,
+  scanIsForSecureSend?: true,
+  transactionData?: TransactionData
 ) {
   let data: { address: string; e164PhoneNumber: string; displayName: string } | undefined
+
   try {
     data = JSON.parse(barcode.data)
   } catch (e) {
     Logger.warn(TAG, 'QR code read failed with ' + e)
   }
   if (typeof data !== 'object' || isEmpty(data.address)) {
-    yield put(showError(ErrorMessages.QR_FAILED_NO_ADDRESS))
+    // yield put(showError(ErrorMessages.QR_FAILED_NO_ADDRESS))
+    yield put(showError(ErrorMessages.QR_FAILED_INVALID_RECIPIENT))
     return
   }
   if (!isValidAddress(data.address)) {
     yield put(showError(ErrorMessages.QR_FAILED_INVALID_ADDRESS))
     return
   }
+  try {
+    if (scanIsForSecureSend) {
+      if (!transactionData) {
+        throw Error(
+          'Error passing through data: Transaction and recipient information not received'
+        )
+      }
+
+      const targetAddress = data.address
+      const { e164PhoneNumber } = transactionData.recipient
+      if (!e164PhoneNumber) {
+        throw Error('Error passing through data: Phone number not part of recipient data')
```

Comment:
Nope, same as above. I can put a comment to make that clear. More of a typescript thing because we use a broad Recipient type that has phone number as optional when in some cases, such as this one, it is actually required.

I was thinking of opening up an issue to do a refactor on that. I've noticed there are lots of places where we do that simple check + error throwing across the codebase. What do you think?